chain/chain/src/chain.rs

```diff
@@ -2485,7 +2485,10 @@ impl<'a> ChainUpdate<'a> {
 
         // Prevent time warp attacks and some timestamp manipulations by forcing strict
         // time progression.
-        if header.inner_lite.timestamp <= prev_header.inner_lite.timestamp {
+        if header.timestamp()
```

Comment:
Why do we add anything here, why `header.timestamp() <= prev_header.timestamp()` is not enough?

Reply:
Because an attacker can produce a lot of valid blocks with 1 nanosecond difference in timestamps.

Reply:
1 millisecond difference would still allow attacker to perform exactly the same attack by producing blocks 1000 timers faster. How are we solving a problem when attacker produces 1000 blocks on their fork while valid block producer creates just one block?

Reply:
That's why we have 300 millisecond difference here. It doesn't completely solve the problem, but help mitigate it.

Reply:
I hope there is a plan for how to solve the problem completely. In my opinion, when it comes to preventing someone from overtaking the blockchain having a hack that makes it only harder is closer to not having a solution than to having a solution. @SkidanovAlex

Reply:
The proper solution is to have a fork choice rule that is resilient to such attack. This check also helps prevent validators from pushing timestamp back too much.